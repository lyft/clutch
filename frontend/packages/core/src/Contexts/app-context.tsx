import React from "react";

import type { Workflow } from "../AppProvider/workflow";

/**
 * Union type representing different lookup keys
 */
export type HeaderItem = "NPS";

interface HeaderItemData {
  /**
   * Optional configuration data to be passed when opening a component
   */
  [key: string]: unknown;
}

export type TriggeredHeaderData = {
  /**
   * The components name, referenced from the above HeaderItems, to be used as a lookup key
   */
  [key in HeaderItem]: HeaderItemData;
};

interface ContextProps {
  workflows: Workflow[];
  /**
   * Will trigger the given header item based on the key, saving any optional data
   * This is useful for when a component would like to trigger an item in the header through some state change,
   * such as the NPS component opening or opening a search and prefilling it
   *
   * Example of utilizing with NPS
   * Opening - triggerHeaderItem("NPS", { defaultOption: "Example 1" })
   * Closing - triggerHeaderItem("NPS", null)
   */
  triggerHeaderItem?: (item: HeaderItem, data?: unknown) => void;
  /**
   * Will return the triggered data, used as a lookup for listening components
   */
  triggeredHeaderData?: TriggeredHeaderData;
}

const ApplicationContext = React.createContext<ContextProps>(undefined);

const useAppContext = () => {
  return React.useContext<ContextProps>(ApplicationContext);
};

export { ApplicationContext, useAppContext };
