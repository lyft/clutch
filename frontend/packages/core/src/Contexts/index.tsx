export { ApplicationContext, useAppContext } from "./app-context";
export { ShortLinkContext, useShortLinkContext } from "./short-link-context";
export { WizardContext, useWizardContext } from "./wizard-context";
export { WorkflowStorageContext, useWorkflowStorageContext } from "./workflow-storage-context";
