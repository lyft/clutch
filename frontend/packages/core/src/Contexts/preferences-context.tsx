import React from "react";
import _ from "lodash";

export type ActionType = "SetPref" | "RemovePref" | "SetLocalPref" | "RemoveLocalPref";
const STORAGE_KEY = "userPreferences";
type State = { key: string; value?: unknown };
type Action = { type: ActionType; payload: State };
type Dispatch = (action: Action) => void;
type UserPreferencesProviderProps = { children: React.ReactNode };
const DEFAULT_PREFERENCES: State = {
  timeFormat: "UTC",
  banners: {
    header: {},
    multiWorkflow: {},
    perWorkflow: {},
  },
} as any;
interface ContextProps {
  preferences: State;
  dispatch: Dispatch;
}

type ContextType = ContextProps | undefined;

export interface PreferencesContextProps {
  preferences: {
    [key: string]: unknown;
  };
  getPref: (key: string) => unknown;
  setPref: (key: string, value: unknown) => void;
}

const preferencesReducer = (preferences: State, action: Action): State => {
  switch (action.type) {
    case "SetPref": {
      const updatedPref = { ...preferences, [action.payload.key]: action.payload.value };
      try {
        localStorage.setItem(STORAGE_KEY, JSON.stringify(updatedPref));
      } catch {}
      return updatedPref;
    }
    case "RemovePref": {
      const updatedPref = { ...preferences };
      delete updatedPref[action.payload.key];
      try {
        localStorage.setItem(STORAGE_KEY, JSON.stringify(updatedPref));
      } catch {}
      return updatedPref;
    }
    case "SetLocalPref": {
      return { ...preferences, [action.payload.key]: action.payload.value };
    }
    case "RemoveLocalPref": {
      const updatedPref = { ...preferences };
      delete updatedPref[action.payload.key];
      return updatedPref;
    }
    default: {
      throw new Error(`Unhandled action type: ${action.type}`);
    }
  }
};

const UserPreferencesContext = React.createContext<ContextType>(undefined);

const UserPreferencesProvider = ({ children }: UserPreferencesProviderProps) => {
  // Load preferences as default value and if none then default to value
  let pref = DEFAULT_PREFERENCES;
  try {
    pref = JSON.parse(localStorage.getItem(STORAGE_KEY) || "");
    // If there are any missing default preferences, add them
    Object.keys(DEFAULT_PREFERENCES).forEach(key => {
      if (_.isEmpty(pref[key]) || !pref[key]) {
        pref[key] = DEFAULT_PREFERENCES[key];
      }
    });
  } catch {
    localStorage.removeItem(STORAGE_KEY);
  }

  localStorage.setItem(STORAGE_KEY, JSON.stringify(pref));
  const [state, dispatch] = React.useReducer(preferencesReducer, pref);

  const value = React.useMemo(() => ({ preferences: state, dispatch }), [state, dispatch]);

  return (
    <UserPreferencesContext.Provider value={value}>{children}</UserPreferencesContext.Provider>
  );
};

const useUserPreferences = () => {
  const context = React.useContext(UserPreferencesContext);
  if (!context) {
    throw new Error("useUserPreferences was invoked outside of a valid context");
  }
  return context;
};

export { useUserPreferences, UserPreferencesProvider };
