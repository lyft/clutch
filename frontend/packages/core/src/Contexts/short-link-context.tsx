import React from "react";

import type { HydratedData } from "./workflow-storage-context/types";

export interface ShortLinkContextProps {
  removeWorkflowSession: () => void;
  retrieveWorkflowSession: () => HydratedData;
  storeWorkflowSession: (data: HydratedData) => void;
}

const ShortLinkContext = React.createContext<ShortLinkContextProps>(undefined);

const useShortLinkContext = () => {
  return React.useContext<ShortLinkContextProps>(ShortLinkContext);
};

export { ShortLinkContext, useShortLinkContext };
