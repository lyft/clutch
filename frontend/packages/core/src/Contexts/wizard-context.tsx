import React from "react";

export interface ContextProps {
  displayWarnings: (warnings: string[]) => void;
  onBack: (params?: { toOrigin?: boolean }) => void;
  onSubmit: () => void;
  setOnSubmit: (f: (...args: any[]) => void) => void;
  setIsLoading: (isLoading: boolean) => void;
  setHasError: (hasError: boolean) => void;
}

const WizardContext = React.createContext<() => ContextProps>(undefined);

const useWizardContext = () => {
  return React.useContext<() => ContextProps>(WizardContext)();
};

export { WizardContext, useWizardContext };
