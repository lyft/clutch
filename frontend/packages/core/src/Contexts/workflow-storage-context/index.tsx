export { WorkflowStorageContext, useWorkflowStorageContext } from "./context";
