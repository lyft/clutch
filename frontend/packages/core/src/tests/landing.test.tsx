import React from "react";
import createRouterContext from "react-router-test-context";
import { shallow } from "enzyme";

import * as appContext from "../Contexts/app-context";
import Landing from "../landing";

describe("Landing component", () => {
  let component;
  const context = createRouterContext();

  beforeAll(() => {
    appContext.useAppContext = jest.fn().mockReturnValue({ workflows: [] });
    component = shallow(<Landing workflows={[]} />, { context });
  });

  it("renders correctly", () => {
    expect(component.debug()).toMatchSnapshot();
  });
});
