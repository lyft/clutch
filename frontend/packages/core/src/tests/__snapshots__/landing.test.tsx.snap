// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Landing component renders correctly 1`] = `
"<Styled(WithStyles(ForwardRef(Paper))) id=\\"landing\\" elevation={0}>
  <WithStyles(ForwardRef(Typography)) variant=\\"h5\\">
    <strong>
      Welcome 
      Anonymous
       
    </strong>
    <span role=\\"img\\" aria-label=\\"Hand Waving\\">
      ðŸ‘‹
    </span>
  </WithStyles(ForwardRef(Typography))>
  <div>
    <WithStyles(ForwardRef(Typography)) gutterBottom={true} variant=\\"body1\\" paragraph={true}>
      Clutch will assist you in safely modifying resources outside of the normal orchestration process.
    </WithStyles(ForwardRef(Typography))>
  </div>
  <styled.div>
    <Styled(WithStyles(ForwardRef(Grid))) container={true} justify=\\"center\\">
      <WithStyles(ForwardRef(Grid)) item={true}>
        <Styled(GitHubIcon) fontSize=\\"small\\" />
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid)) item={true}>
        <Styled(WithStyles(ForwardRef(Link))) target=\\"_blank\\" rel=\\"noreferrer\\" href=\\"https://github.com/lyft/clutch\\" underline=\\"none\\">
          lyft/clutch
        </Styled(WithStyles(ForwardRef(Link)))>
      </WithStyles(ForwardRef(Grid))>
    </Styled(WithStyles(ForwardRef(Grid)))>
  </styled.div>
</Styled(WithStyles(ForwardRef(Paper)))>"
`;
